tests/test_wsgi.py

```diff
@@ -29,38 +29,454 @@
+    wsgi_env['start_time'] = time.time() - 1.0
+    response = wsgi.WSGIResponse(wsgi_env, start_response)
+    response.start_response('200 OK', [], None)
+    output = b''.join(response.wrap([b'output', b' ', b'here']))
+
+    assert output == b'output here'
+    context.assert_log(
+        msg='GET /',
+        extra=dict([
+            ('method', 'GET'),
+            ('path', '/'),
+            ('status', 200),
+            ('duration_ms', 1000.0),
+            ('ip', '127.0.0.1'),
+            ('proto', 'HTTP/1.0'),
+            ('length', len(output)),
+        ]),
+    )
+
+
+@freeze_time('2016-01-02 03:04:05.1234')
+def test_wsgi_response_wrap_file(wsgi_env, start_response, context, tmpdir):
+    path = tmpdir.join('filecontent')
+    path.write('CONTENT')
+    wsgi_env['start_time'] = time.time() - 1.0
+    wsgi_env['wsgi.file_wrapper'] = wsgiref.util.FileWrapper
+
+    response = wsgi.WSGIResponse(wsgi_env, start_response)
+    response.start_response('200 OK', [], None)
+    wrapper = wsgiref.util.FileWrapper(open(str(path)))
+    respiter = response.wrap(wrapper)
+    output = ''.join(respiter)
+    respiter.close()
+
+    assert output == 'CONTENT'
+    context.assert_log(
+        msg='GET /',
+        extra=dict([
+            ('method', 'GET'),
+            ('path', '/'),
+            ('status', 200),
+            ('duration_ms', 1000.0),
+            ('ip', '127.0.0.1'),
+            ('proto', 'HTTP/1.0'),
+            ('length', len(output)),
+            ('filepath', str(path)),
+        ]),
+    )
+
+
+@freeze_time('2016-01-02 03:04:05.1234')
+def test_wsgi_response_wrap_error(wsgi_env, start_response, context):
+    wsgi_env['start_time'] = time.time() - 1.0
+    response = wsgi.WSGIResponse(wsgi_env, start_response)
+    response.start_response('200 OK', [], None)
+
+    class ErrorGenerator():
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            raise Exception('error')
+
+    output = b''.join(response.wrap(ErrorGenerator()))
+
+    assert output == b'Exception'
+
+    context.assert_log(
+        msg='GET /',
+        extra=dict([
+            ('method', 'GET'),
+            ('path', '/'),
+            ('status', 500),
+            ('duration_ms', 1000.0),
+            ('ip', '127.0.0.1'),
+            ('proto', 'HTTP/1.0'),
+            ('length', len(output)),
+            ('exc_type', 'Exception'),
+        ]),
+    )
+
+
+@freeze_time('2016-01-02 03:04:05.1234')
+def test_wsgi_response_wrap_error_headers_sent(
+        wsgi_env, start_response, context):
+    wsgi_env['start_time'] = time.time() - 1.0
+    response = wsgi.WSGIResponse(wsgi_env, start_response)
+    response.start_response('200 OK', [], None)
+
+    def iterator():
+        start_response.headers_sent = True
+        yield b'some content'
+        raise Exception('error')
+
+    it = response.wrap(iterator())
+    with pytest.raises(Exception):
+        list(it)
```

Comment:
Am I missing something here? Where's the assert that `start_response.headers_sent` is `True`?

Reply:
headers_sent is part of the setup, not the assertion. By setting that, we alter the behaviour of mock_send_response to simulate the case when headers have been sent, in which case, the WSGI spec says the WSGI server should reraise the exc.